typing/typecore.ml

```diff
@@ -3740,18 +3757,20 @@ and type_expect_
        [Nolabel, sbody]) ->
       if not (Clflags.Extension.is_enabled Local) then
         raise (Typetexp.Error (loc, Env.empty, Local_not_enabled));
-      submode ~loc ~env Value_mode.local expected_mode;
+      let mode = expect_mode_cross env ty_expected mode_local in
+      submode ~loc ~env mode.mode expected_mode;
       let exp =
-        type_expect ?in_function ~recarg env mode_local sbody
+        type_expect ?in_function ~recarg env mode sbody
           ty_expected_explained
       in
       { exp with exp_loc = loc }
   | Pexp_apply
       ({ pexp_desc = Pexp_extension({txt = ("ocaml.local" | "local")}, PStr []) },
        [Nolabel, sbody]) ->
-      submode ~loc ~env Value_mode.local expected_mode;
+      let mode = expect_mode_cross env ty_expected mode_local in
```

Comment:
I don't understand what this case is for - when we're already at `mode_local`, what does mode crossing do?

Reply:
My understanding is that, the original code enforces `local_ 42` to be local. With this modification, however, `local_ 42` would be unconstrained (and could be global)

Reply:
Thanks for the explanation! I now see that the change accepts more programs and is sound, but I'm still not sure I actually want it - it seems confusing that `ref (local_ 42)` is accepted. Is there a use-case I'm missing here?

Reply:
Yeah Leo and I wasn't sure either. But if we stick to the principle that "immediates don't respect mode", we should probabaly allow `ref (local_ 42)` to type check?

Reply:
As something of an outside observer, I'd expect `ref (local_ 42)` to be just fine: immediates simply don't care about locality. That seems simpler than trying (somehow) to avoid this.

Reply:
OK, let's go with this as is!